File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_date_range/DateRangePickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogCallbacks;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom.CustomDialogBuilder;

import java.time.OffsetDateTime;
import java.util.Date;

import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LifecycleOwner;

public class DateRangePickerDialogBuilder extends BaseDialogBuilder<DateRangePickerDialog> {

    private BasePickerDialogFragment.SelectionChangedListener<OffsetDateTime[]> dateSetListener;
    private String textFrom;
    private String textTo;
    private String dateFormat;
    private OffsetDateTime[] initialRange;

    public DateRangePickerDialogBuilder(FragmentManager manager, String dialogTag) {
        super(manager, dialogTag);
    }

    public DateRangePickerDialogBuilder(LifecycleOwner lifecycleOwner, FragmentManager manager, String dialogTag) {
        super(lifecycleOwner, manager, dialogTag);
    }

    @Override
    public DateRangePickerDialogBuilder withTitle(String title) {
        return (DateRangePickerDialogBuilder) super.withTitle(title);
    }

    @Override
    public DateRangePickerDialogBuilder withMessage(String message) {
        return (DateRangePickerDialogBuilder) super.withMessage(message);
    }

    @Override
    public DateRangePickerDialogBuilder withPositiveButton(BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (DateRangePickerDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public DateRangePickerDialogBuilder withNegativeButton(BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (DateRangePickerDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public DateRangePickerDialogBuilder withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        return (DateRangePickerDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public DateRangePickerDialogBuilder withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        return (DateRangePickerDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public DateRangePickerDialogBuilder withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        return (DateRangePickerDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public DateRangePickerDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (DateRangePickerDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public DateRangePickerDialogBuilder withDialogType(BaseDialogFragment.DialogTypes dialogType) {
        return (DateRangePickerDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public DateRangePickerDialogBuilder withAnimations(BaseDialogFragment.DialogAnimationTypes animation) {
        return (DateRangePickerDialogBuilder) super.withAnimations(animation);
    }

    @Override
    public DateRangePickerDialogBuilder withDialogCallbacks(DialogCallbacks callbacks) {
        return (DateRangePickerDialogBuilder) super.withDialogCallbacks(callbacks);
    }

    public DateRangePickerDialogBuilder withTextFrom(String textFrom) {
        this.textFrom = textFrom;
        return this;
    }

    public DateRangePickerDialogBuilder withTextTo(String textTo) {
        this.textTo = textTo;
        return this;
    }

    public DateRangePickerDialogBuilder withOnDateSelectedEvent(BasePickerDialogFragment.SelectionChangedListener<OffsetDateTime[]> listener) {
        this.dateSetListener = listener;
        return this;
    }

    public DateRangePickerDialogBuilder withSelection(OffsetDateTime[] range) {
        this.initialRange = range;
        return this;
    }

    public DateRangePickerDialogBuilder withDateFormat(String dateFormat) {
        this.dateFormat = dateFormat;
        return this;
    }

    @Override
    public DateRangePickerDialog buildDialog() {
        DateRangePickerDialog dialogFragment = (DateRangePickerDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = DateRangePickerDialog.newInstance(positiveButtonConfiguration, negativeButtonConfiguration, dateFormat, textFrom, textTo, cancelableOnClickOutside, animation);
        dialogFragment.setLifecycleOwner(dialogLifecycleOwner);
        dialogFragment.setDialogCallbacks(dialogCallbacks);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener);
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener);
        dialogFragment.setOnSelectionChangedListener(dateSetListener);
        dialogFragment.setSelection(initialRange);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_date_range/DateRangePickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_date_range

import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogButtonConfiguration
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogButtonClickListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogShowListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogHideListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogCancelListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogAnimationTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogCallbacks
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import androidx.fragment.app.FragmentManager
import java.time.OffsetDateTime

class DateRangePickerDialogBuilder : BaseDialogBuilder<DateRangePickerDialog?> {
    private var dateSetListener: SelectionChangedListener<Array<OffsetDateTime>>? = null
    private var textFrom: String? = null
    private var textTo: String? = null
    private var dateFormat: String? = null
    private var initialRange: Array<OffsetDateTime?>

    constructor(manager: FragmentManager?, dialogTag: String?) : super(manager, dialogTag) {}
    constructor(lifecycleOwner: LifecycleOwner?, manager: FragmentManager?, dialogTag: String?) : super(
        lifecycleOwner,
        manager,
        dialogTag
    ) {
    }

    override fun withTitle(title: String?): DateRangePickerDialogBuilder? {
        return super.withTitle(title) as DateRangePickerDialogBuilder
    }

    override fun withMessage(message: String?): DateRangePickerDialogBuilder? {
        return super.withMessage(message) as DateRangePickerDialogBuilder
    }

    override fun withPositiveButton(
        configuration: DialogButtonConfiguration?,
        onClickListener: DialogButtonClickListener?
    ): DateRangePickerDialogBuilder? {
        return super.withPositiveButton(configuration, onClickListener) as DateRangePickerDialogBuilder
    }

    override fun withNegativeButton(
        configuration: DialogButtonConfiguration?,
        onClickListener: DialogButtonClickListener?
    ): DateRangePickerDialogBuilder? {
        return super.withNegativeButton(configuration, onClickListener) as DateRangePickerDialogBuilder
    }

    override fun withOnCancelListener(listener: DialogCancelListener?): DateRangePickerDialogBuilder? {
        return super.withOnCancelListener(listener) as DateRangePickerDialogBuilder
    }

    override fun withOnShowListener(listener: DialogShowListener?): DateRangePickerDialogBuilder? {
        return super.withOnShowListener(listener) as DateRangePickerDialogBuilder
    }

    override fun withOnHideListener(listener: DialogHideListener?): DateRangePickerDialogBuilder? {
        return super.withOnHideListener(listener) as DateRangePickerDialogBuilder
    }

    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): DateRangePickerDialogBuilder? {
        return super.withCancelOnClickOutside(closeOnClickOutside) as DateRangePickerDialogBuilder
    }

    override fun withDialogType(dialogType: DialogTypes?): DateRangePickerDialogBuilder? {
        return super.withDialogType(dialogType) as DateRangePickerDialogBuilder
    }

    override fun withAnimations(animation: DialogAnimationTypes?): DateRangePickerDialogBuilder? {
        return super.withAnimations(animation) as DateRangePickerDialogBuilder
    }

    override fun withDialogListeners(callbacks: DialogCallbacks?): DateRangePickerDialogBuilder? {
        return super.withDialogListeners(callbacks) as DateRangePickerDialogBuilder
    }

    fun withTextFrom(textFrom: String?): DateRangePickerDialogBuilder {
        this.textFrom = textFrom
        return this
    }

    fun withTextTo(textTo: String?): DateRangePickerDialogBuilder {
        this.textTo = textTo
        return this
    }

    fun withOnDateSelectedEvent(listener: SelectionChangedListener<Array<OffsetDateTime>>?): DateRangePickerDialogBuilder {
        dateSetListener = listener
        return this
    }

    fun withSelection(range: Array<OffsetDateTime?>): DateRangePickerDialogBuilder {
        initialRange = range
        return this
    }

    fun withDateFormat(dateFormat: String?): DateRangePickerDialogBuilder {
        this.dateFormat = dateFormat
        return this
    }

    override fun buildDialog(): DateRangePickerDialog? {
        var dialogFragment = fragmentManager!!.findFragmentByTag(dialogTag) as DateRangePickerDialog?
        if (dialogFragment == null) dialogFragment = DateRangePickerDialog.Companion.newInstance(
            positiveButtonConfiguration,
            negativeButtonConfiguration,
            dateFormat,
            textFrom,
            textTo,
            cancelableOnClickOutside,
            animation
        )
        dialogFragment.setLifecycleOwner(dialogLifecycleOwner)
        dialogFragment.setDialogCallbacks(dialogListeners)
        dialogFragment.setFragmentManager(fragmentManager)
        dialogFragment.setDialogTag(dialogTag)
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener)
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener)
        dialogFragment.setOnSelectionChangedListener(dateSetListener)
        dialogFragment.selection = initialRange
        dialogFragment.addOnShowListener(onShowListener)
        dialogFragment.addOnHideListener(onHideListener)
        dialogFragment.addOnCancelListener(onCancelListener)
        return dialogFragment
    }
}